xbmc/interfaces/json-rpc/PlaylistOperations.cpp

```diff
@@ -212,6 +212,87 @@ JSONRPC_STATUS CPlaylistOperations::Swap(const std::string &method, ITransportLa
   return ACK;
 }
 
+JSONRPC_STATUS CPlaylistOperations::SetShuffle(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
+{
+  int playlist = GetPlaylist(parameterObject["playlistid"]);
+  CVariant shuffle = parameterObject["shuffle"];
+  CGUIWindowSlideShow *slideshow = NULL;
+  switch (playlist)
+  {
+    case PLAYLIST_MUSIC:
+    case PLAYLIST_VIDEO:
+    {
+      if (CServiceBroker::GetPlaylistPlayer().IsShuffled(playlist))
+      {
+        if ((shuffle.isBoolean() && !shuffle.asBoolean()) ||
+            (shuffle.isString() && shuffle.asString() == "toggle"))
+        {
+          CApplicationMessenger::GetInstance().SendMsg(TMSG_PLAYLISTPLAYER_SHUFFLE, playlist, 0);
+        }
+      }
+      else
+      {
+        if ((shuffle.isBoolean() && shuffle.asBoolean()) ||
+            (shuffle.isString() && shuffle.asString() == "toggle"))
+        {
+          CApplicationMessenger::GetInstance().SendMsg(TMSG_PLAYLISTPLAYER_SHUFFLE, playlist, 1);
+        }
+      }
+      break;
+    }
+
+    case PLAYLIST_PICTURE:
+      slideshow = CServiceBroker::GetGUI()->GetWindowManager().GetWindow<CGUIWindowSlideShow>(WINDOW_SLIDESHOW);
+      if (!slideshow)
+        return FailedToExecute;
+      if (slideshow->IsShuffled())
+      {
+        if ((shuffle.isBoolean() && !shuffle.asBoolean()) ||
+            (shuffle.isString() && shuffle.asString() == "toggle"))
+          return FailedToExecute;
```

Comment:
So there is no way to "un-shuffle" a slideshow?

Reply:
I used the code for the [`CPlayerOperations::SetShuffle`](https://github.com/xbmc/xbmc/blob/21a838b61d4d0df7ceb020a2a10a67c233ed73f4/xbmc/interfaces/json-rpc/PlayerOperations.cpp#L959) method. Looking in the history, it looks like we cannot _unshuffle_ a slideshow since commit https://github.com/xbmc/xbmc/commit/c4f161e177f5e079f2ba67c264c2ebb85a9a2b03:

- Before:

```cpp
JSON_STATUS CPlaylistOperations::UnShuffle(const CStdString &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
{
  CSingleLock lock(VirtualCriticalSection);
  CPlayListPtr playlist = GetPlaylist(parameterObject);

  if (playlist)
  {
    playlist->UnShuffle();
    return ACK;
  }

  return InvalidParams;
}
```

- After:

```cpp
JSON_STATUS CPlayerOperations::UnShuffle(const CStdString &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
{
  switch (GetPlayer(parameterObject["playerid"]))
  {
    case Video:
    case Audio:
      g_application.getApplicationMessenger().PlayListPlayerShuffle(GetPlaylist(GetPlayer(parameterObject["playerid"])), false);
      OnPlaylistChanged();
      break;

    case Picture:
    default:
      return FailedToExecute;
  }
  return ACK;
}
```